### ISSUE TITLE
grep hangs with: zsh -ixc : 2>&1 | grep p10k

### ISSUE THREAD

**Issue by apolopena** [romkatv/powerlevel10k#737]
When I use the **powerlevel10k** theme and I then run the command `zsh -ixc : 2>&1 | grep p10k` grep hangs after all the matches are found, forcing me to exit with ctrl+c. 

`grep` exits cleanly when I use other **oh-my-zsh** themes such as `robbyrussell`.

To reproduce (in `.zshrc`):
`ZSH_THEME="powerlevel10k/powerlevel10k"`
reload shell, run the command `zsh -ixc : 2>&1 | grep p10k`
**Result**: `grep` hangs after matches are displayed
**Expected**: `grep` exits cleanly.

Then (in `.zshrc`) change theme to `robbyrussell`
`ZSH_THEME="robbyrussell"`
reload shell, run the command `zsh -ixc : 2>&1 | grep p10k`
**Result**: `grep` exits cleanly
**Expected**: `grep` exits cleanly.

**Comment by romkatv** [comment]
Cannot reproduce following the provided instructions. Can you?

1. Run this command:
```zsh
(
  emulate -L zsh -o err_return -o pushd_silent
  local tmp
  tmp="$(mktemp -d ${TMPDIR:-/tmp}/p10k.XXXXXXXXXX)"
  {
    builtin pushd -q -- $tmp
    ${=${commands[curl]:+curl -fsSLO}:-wget} \
      https://raw.githubusercontent.com/ohmyzsh/ohmyzsh/master/tools/install.sh
    export ZSH=$tmp/.oh-my-zsh
    sh ./install.sh --unattended --keep-zshrc
    git clone --depth=1 https://github.com/romkatv/powerlevel10k.git .oh-my-zsh/custom/themes/powerlevel10k
    command cp .oh-my-zsh/templates/zshrc.zsh-template .zshrc
    sed -i.bak 's|robbyrussell|powerlevel10k/powerlevel10k|' .zshrc
    sed -i.bak '/^export ZSH=/d' .zshrc
    print 'POWERLEVEL9K_DISABLE_CONFIGURATION_WIZARD=true' >>.zshrc
    builtin popd -q
    ZDOTDIR=$tmp zsh
  } always {
    cd /
    rm -rf -- $tmp
  }
)
```
2. When prompt appears, run this command:
```zsh
zsh -ixc : 2>&1 | grep p10k
```

Does the second command hang?

**Comment by apolopena** [comment]
I ran the command, 'grep' still hangs the matches are displayed.

**Comment by romkatv** [comment]
What is the output of the following command?

```zsh
() {
  emulate -L zsh -o xtrace
  uname -a
  typeset -p 'ZSH_VERSION|ZSH_PATCHLEVEL'
}
```

Can you tell me anything else about this? Have you tried changing the command? Have you discovered what makes it stop hanging? For example, is `:` necessary or will hand with `''` too? Is it necessary to call `grep`? Will it hang with `true` instead? Is it necessary to pipe? Is it necessary to redirect stderr to stdout?

**Comment by apolopena** [comment]
I just ran the your test command again in an asciinema session and oddly the `zsh -ixc : 2>&1 | grep p10k` exited gracefully. 
I will test a bit more and get back to you shortly.

> _Have you tried changing the command?_
- I have tried other variations of the command with the same hung result such as replacing `grep` with `wc`.
> _Have you discovered what makes it stop hanging?_
- No but I am close.
>  _Is `:` necessary or will `grep` hang with `''` too? Will it hang with true instead?_
- `:` `''` and `true` in the command all render the same hung result
> _Is it necessary to call grep?_
- Yes
> _Is it necessary to pipe?_
- I am sure it could be done another way but I prefer to pipe it because that is whay I know. I am open to other ways though. I am just learning shell scripting.
> _Is it necessary to redirect stderr to stdout?_
- Yes

**Comment by romkatv** [comment]
> What is the output of the following command?
> 
> ```shell
> () {
>   emulate -L zsh -o xtrace
>   uname -a
>   typeset -p 'ZSH_VERSION|ZSH_PATCHLEVEL'
> }
> ```

Please :point_up:

**Comment by apolopena** [comment]
There were errors in the response to that command (`(anon):typeset:3: no such variable: ZSH_VERSION|ZSH_PATCHLEVEL`),  but here is the info you wanted :
```
Linux DELL9550 4.4.0-18362-Microsoft #836-Microsoft Mon May 05 16:04:00 PST 2020 x86_64 x86_64 x86_64 GNU/Linux
ubuntu/5.8-3ubuntu1
```

**Comment by apolopena** [comment]
I noticed that when running in the testing subshell (your test command), it appears that if the results are long enough (lots of matches) then grep hangs. If the results are shorter (less matches) or there are no matches then it doesn't hang.

**Comment by romkatv** [comment]
Isn’t the output always the same? How can there be fewer or more matches on different runs?

**Comment by romkatv** [comment]
Could you run this and post the output?

    typeset -pm 'ZSH_VERSION|ZSH_PATCHLEVEL'

The command I posted earlier was missing `-m`.

**Comment by apolopena** [comment]
Result of `typeset -pm 'ZSH_VERSION|ZSH_PATCHLEVEL'`
```
typeset ZSH_PATCHLEVEL=ubuntu/5.8-3ubuntu1
typeset ZSH_VERSION=5.8
```

> _Isn’t the output always the same? How can there be fewer or more matches on different runs?_
- I did some alternate testing by trying some subsequent runs with grep commands that I know would return a shorter result such as `zsh -ixc : 2>&1 | grep 'preexec periodic zshaddhistory'`
- To clarify: `zsh -ixc : 2>&1 | grep 'preexec periodic zshaddhistory'` exits gracefully in the testing subshell and hangs when run in my standard environment.

**Comment by tmpm697** [comment]
I don't have this issue, `grep` exits cleanly.

**Comment by romkatv** [comment]
@apolopena Please try the following.

1. Run this command:
```zsh
(
  emulate -L zsh -o err_return -o pushd_silent
  local tmp
  tmp="$(mktemp -d ${TMPDIR:-/tmp}/p10k.XXXXXXXXXX)"
  {
    builtin pushd -q -- $tmp
    ${=${commands[curl]:+curl -fsSLO}:-wget} \
      https://raw.githubusercontent.com/ohmyzsh/ohmyzsh/master/tools/install.sh
    export ZSH=$tmp/.oh-my-zsh
    sh ./install.sh --unattended --keep-zshrc
    command cp .oh-my-zsh/templates/zshrc.zsh-template .zshrc
    sed -i.bak '/^export ZSH=/d' .zshrc
    >>.zshrc <<\END
() {
  emulate -L zsh
  local i
  for i in {1..1000000}; do
    : p10k $i
  done
}
END
    builtin popd -q
    ZDOTDIR=$tmp zsh
  } always {
    cd /
    rm -rf -- $tmp
  }
)
```
2. When prompt appears, run this command:
```zsh
zsh -ixc : 2>&1 | grep p10k
```

Does the second command hang?

**Comment by apolopena** [comment]
@romkatv Ok I did and the second command does not hang.

**Comment by apolopena** [comment]
I am beginning to suspect these large files in my `/tmp` could be the culprit. My system won't or hasn't cleaned them up. I have no `systemd` on WSL1.
```
-rw-rw-rw- 1 apena apena    678344 May 18 12:22 p10k-instant-prompt-output-apena-15064
-rw-rw-rw- 1 apena apena    678344 May 18 12:23 p10k-instant-prompt-output-apena-15292
-rw-rw-rw- 1 apena apena   7636122 May  2 21:09 p10k-instant-prompt-output-apena-5342
-rw-rw-rw- 1 apena apena    462090 May  3 10:50 p10k-instant-prompt-output-apena-6593
-rw-rw-rw- 1 apena apena 739135884 May  3 11:35 p10k-instant-prompt-output-apena-6606
-rw-rw-rw- 1 apena apena 219125586 May  3 11:10 p10k-instant-prompt-output-apena-6621
-rw-rw-rw- 1 apena apena   7181801 May  3 11:35 p10k-instant-prompt-output-apena-6636
-rw-rw-rw- 1 apena apena  20664273 May  3 11:36 p10k-instant-prompt-output-apena-6648
-rw-rw-rw- 1 apena apena   1693813 May  3 11:36 p10k-instant-prompt-output-apena-6661
-rw-rw-rw- 1 apena apena   3000638 May  3 11:36 p10k-instant-prompt-output-apena-6673
-rw-rw-rw- 1 apena apena         0 May 14 13:14 p10k-instant-prompt-output-apena-7
```
I will delete them manually and test again. @romkatv do you know what makes these tmp files so large?

**Comment by romkatv** [comment]
> I am beginning to suspect these large files in my `/tmp` could be the culprit.

That's unlikely.

> My system won't or hasn't cleaned them up. I have no `systemd` on WSL1.

It's a good idea to clear `/tmp` in WSL1 after reboot.

> @romkatv do you know what makes these `p10k-instant-prompt-output....` files so large?

You'll have the answer if you look at what these files contain.

May I ask why you run `zsh -ixc : 2>&1 | grep p10k` in the first place?

**Comment by apolopena** [comment]
No it wasnt that.

**Comment by apolopena** [comment]
The tmp files were large because they were caching the results of the command I kept testing.

> _May I ask why you run zsh -ixc : 2>&1 | grep p10k in the first place?_
- Actually grepping for `p10k` is arbitrary. I use `zsh -ixc : 2>&1 | grep ANYSTRING` to help locate the origin of aliases in my shell. I was tracking down a bug from the **oh-my-zsh** `colorize` plugin ([ccat alias](https://github.com/ohmyzsh/ohmyzsh/issues/8946#issuecomment-631416924)) and I wanted to make the command an autoloaded function but it hung on me so I poked around and noticed I could make it stop hanging by not using the p10k theme so I logged a bug here. So it is not super important that I resolve this issue with my environment and p10k. I am just very curious at this point, it's learning all for me and I guess I would like to use that command when I am using the p10k theme in the future but it is not imperative.
